app/src/main/java/com/github/gotify/messages/provider/MessageStateHolder.java

```diff
@@ -66,40 +77,48 @@ synchronized int getLastReceivedMessage() {
         return lastReceivedMessage;
     }
 
-    synchronized void removeMessage(Message message) {
+    synchronized void deleteMessage(Message message) {
         MessageState allMessages = state(MessageState.ALL_MESSAGES);
         MessageState appMessages = state(message.getAppid());
 
+        int pendingDeletedAllPosition = -1;
+        int pendingDeletedAppPosition = -1;
+
         if (allMessages.loaded) {
             int allPosition = allMessages.messages.indexOf(message);
             allMessages.messages.remove(allPosition);
-            lastRemovedAllPosition = allPosition;
+            pendingDeletedAllPosition = allPosition;
         }
 
         if (appMessages.loaded) {
             int appPosition = appMessages.messages.indexOf(message);
             appMessages.messages.remove(appPosition);
-            lastRemovedAppPosition = appPosition;
+            pendingDeletedAppPosition = appPosition;
         }
 
-        lastRemovedMessage = message;
+        pendingDeletion =
+                new MessageDeletion(message, pendingDeletedAllPosition, pendingDeletedAppPosition);
     }
 
-    synchronized PositionPair undoLastRemoveMessage() {
-        PositionPair result = null;
-
-        if (lastRemovedMessage != null) {
-            addMessage(lastRemovedMessage, lastRemovedAllPosition, lastRemovedAppPosition);
-            result = new PositionPair(lastRemovedAllPosition, lastRemovedAppPosition);
-
-            lastRemovedMessage = null;
-            lastRemovedAllPosition = -1;
-            lastRemovedAppPosition = -1;
-        }
+    synchronized MessageDeletion undoPendingDeletion() {
+        if (pendingDeletion != null)
+            addMessage(
+                    pendingDeletion.getMessage(),
+                    pendingDeletion.getAllPosition(),
+                    pendingDeletion.getAppPosition());
+        return purgePendingDeletion();
+    }
 
+    synchronized MessageDeletion purgePendingDeletion() {
+        MessageDeletion result = pendingDeletion;
+        pendingDeletion = null;
         return result;
     }
 
+    boolean deletionPending() {
```

Comment:
This should be synchronized too right? Could you make every non private method in Facade and State synchronized?

Reply:
Done